contributors/design-proposals/storage/volume-topology-scheduling.md

```diff
@@ -78,16 +78,139 @@ pod placement, volume binding and provisioning should be more tightly coupled wi
+object will be extended with a new NodeAffinity field that specifies the
+constraints.  It will closely mirror the existing NodeAffinity type used by
+Pods, but we will use a new type so that we will not be bound by existing and
+future Pod NodeAffinity semantics.
 
+```
+type PersistentVolumeSpec struct {
+    ...
+
+    NodeAffinity *VolumeNodeAffinity
+}
+
+type VolumeNodeAffinity struct {
+    // The PersistentVolume can only be accessed by Nodes that meet
+    // these required constraints
+    Required *NodeSelector
+}
+```
+
+The `Required` field is a hard constraint and indicates that the PersistentVolume
+can only be accessed from Nodes that satisfy the NodeSelector.
+
+In the future, a `Preferred` field can be added to handle soft node contraints with
+weights, but will not be included in the initial implementation.
+
+The advantages of this NodeAffinity field vs the existing method of using zone labels
+on the PV are:
+* We don't need to expose first-class labels for every topology domain.
+* Implementation does not need to be updated every time a new topology domain
+  is added to the cluster.
+* NodeSelector is able to express more complex topology with ANDs and ORs.
+
+Some downsides include:
+* You can have a proliferation of Node labels if you are running many different
+  kinds of volume plugins, each with their own topology labeling scheme.
+
+
+### Example PVs with NodeAffinity
+#### Local Volume
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: local-volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: my-class
+  local:
+    path: /mnt/disks/ssd1
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: kubernetes.io/hostname
+          operator: In
+          values:
+          - node-1
+```
+
+#### Zonal Volume
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: zonal-volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: my-class
+  gcePersistentDisk:
+    diskName: my-disk
+    fsType: ext4
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: failure-domain.beta.kubernetes.io/zone
+          operator: In
+          values:
+          - us-central1-a
+```
+
+#### Multi-Zonal Volume
+```
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  Name: multi-zonal-volume-1
+spec:
+  capacity:
+    storage: 100Gi
+  storageClassName: my-class
+  gcePersistentDisk:
+    diskName: my-disk
+    fsType: ext4
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: failure-domain.beta.kubernetes.io/zone
+          operator: In
+          values:
+          - us-central1-a
+          - us-central1-b
+```
+
+### Bound PVC Enforcement
+For PVCs that are already bound to a PV with NodeAffinity, enforcement is
+simple and will be done at two places:
+* Scheduler predicate: if a Pod references a PVC that is bound to a PV with
```

Comment:
How is it / will it be related to existing predicate that evaluates hard-coded `failure-domain.beta.kubernetes.io/zone`? I assume that both old and new will be enabled then PVC.NodeAffinity feature gate is enabled. If the old predicate is going away, we need some admission plugin that automatically fills PVC.NodeAffinity on upgrade.

Reply:
Ah good point.  They both will have to coexist until this becomes GA + deprecation period, but eventually we should remove the old zone predicate.

Can an admission plugin upgrade existing objects that don't get updated by the user?  Maybe we could do something similar like the pvc/pv protection controller and fill in the PV.NodeAffinity.

Reply:
Added a section on defaults above